### PR TITLE
evo2 SAE eval: label producers + probing harness (on #1629)

### PR DESCRIPTION
**Pull request by polinabinder1**
## Summary
Evo2 SAE **eval: label producers + probing harness** — turn DNA into an `ActivationBuffer` (the one model-touching step) and score it through the probing CLI (AUROC / annotate / linear / domain-F1 / loss-recovered). Lives in the **`evo2_sae.eval.probing`** package, alongside the #1629 primitives.

**Stacked on #1629** (→ #1622). #1630 supplies the eval labels.

## Contents — `evo2_sae.eval.probing`
- `evo2_buffer.py` — DNA → `ActivationBuffer` (the only model-touching code: Evo2 → layer-L residual → `SAE.encode` + per-token labels)
- `labelers.py` — per-token biological labelers (genetic code / CDS frame; prokaryotic gene calling via `pyrodigal`)
- `annot_tracks.py` — BED/GFF interval-track loader → per-token masks (RefSeq / Rfam / JASPAR / ENCODE)
- `euk_windows.py` — eukaryotic gene-structure windows
- `probe.py` — the probing CLI (`extract` / `auroc` / `annotate` / `linear` / `euk-f1` / `domain-eval`)
- `probe_loss_recovered.py` — SAE fidelity (loss recovered); reuses the shared `sae.eval.loss_recovered`

Imports are package-relative; the primitives come from `evo2_sae.eval.probing`. `loss_recovered` stays in the shared `sae` lib (used by esm2/codonfm too).

## How to run
```bash
cd interpretability/sparse_autoencoders/recipes/evo2
bash .ci_build.sh && source .ci_test_env.sh        # or: PYTHONPATH=src:../../sae/src
pytest tests/test_probe_integration.py tests/test_labelers.py tests/test_annot_tracks.py tests/test_euk_windows.py
# CLI: python -m evo2_sae.eval.probing.probe extract|auroc|annotate|linear|domain-eval ...
```
No dedicated CI lane (deferred — see #1622; CI should fold into the repo-wide recipe lane later).

## Tests
- **CPU (no model):** label producers (`labelers` / `annot_tracks` / `euk_windows`) + the probe-CLI integration (buffer save/load roundtrip incl. the dense twin, AUROC/annotate/linear over a planted feature, `domain_f1` over interval tracks). **35 passed.**
- **GPU:** the real-engine buffer/loss-recovered path is gated by `@pytest.mark.skipif(not torch.cuda.is_available())` — runs on a GPU box, skips otherwise.

### CODE REVIEW

**Comment by polinabinder1** [comment]
@coderabbitai review

**Comment by polinabinder1** [comment]
@coderabbitai review

File: interpretability/sparse_autoencoders/recipes/evo2/src/evo2_sae/eval/probing/probe_loss_recovered.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    pairs = sample_sequences(
+        args.fasta, args.n_seqs * args.seq_len, args.seq_len, kingdoms=["prok", "euk"], seed=args.seed
+    )[: args.n_seqs]
+    batches = []
+    for kingdom, dna in pairs:
+        ids = engine.tokenize(KINGDOM_TAGS[kingdom] + dna)
+        if len(ids) > 4:
+            batches.append(torch.tensor([ids], dtype=torch.long, device=dev))
+
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply interpretability/sparse_autoencoders/recipes/evo2/src/evo2_sae/eval/probing/probe_loss_recovered.py:109]: Done in `942071d8` — raises `SystemExit` when no sequences survive the >4-token filter.

File: bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+  probe.py extract       --out BUF [...]        build an ActivationBuffer (needs the model)
+  probe.py auroc         --acts BUF --labels .. per-feature AUROC table (prints)
+  probe.py annotate      --acts BUF --out P     assign each feature its best concept -> annotation parquet
+  probe.py linear        --acts BUF --labels .. SAE-vs-dense single + multi (disentanglement/distributed)
+  probe.py codon-aa      --acts CODON_BUF       codon/AA decoders + family-disjoint, SAE vs dense
+  probe.py euk-f1        --fasta .. --gff ..    RefSeq gene-structure domain-F1 (needs the model)
+  probe.py domain-eval   --fasta .. --track ..  user annotated dataset -> per-feature domain-F1 + AUROC vs
+                                                any BED/GFF tracks (RefSeq/Rfam/JASPAR/ENCODE) (needs the model)
+  probe.py loss-recovered [...]                 fidelity via sae.eval.loss_recovered (needs the model)
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py]: Done in `942071d8` — help no longer lists `loss-recovered` as a subcommand; it points at the separate `probe_loss_recovered.py` script.

File: bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    # Standardize X by the train-split mean/std (reuses sae.eval.probing.standardize).
+    mu, sd = standardize(X, tr)
+    return (X - mu) / sd
+
+
+def _load(a):
+    """Load the probing buffer + resolve requested label names (default: all labels in the buffer)."""
+    buf = ActivationBuffer.load(a.acts)
+    labels = a.labels.split(",") if getattr(a, "labels", None) else list(buf.label_names)
+    return buf, [t for t in labels if t in buf.name_idx]
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py]: Done in `942071d8` — `_load` now raises on unknown `--labels` (lists the buffer's available labels) instead of silently dropping them.

File: interpretability/sparse_autoencoders/recipes/evo2/src/evo2_sae/eval/probing/probe.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    from sae.eval.probing import fit_logreg
+
+    for n in names:
+        ytr = torch.from_numpy(buf.labels[tr.numpy(), buf.name_idx[n]]).to(dev).float()
+        yte = torch.from_numpy(buf.labels[te.numpy(), buf.name_idx[n]]).to(dev)
+        if ytr.sum() in (0, len(ytr)) or yte.sum() == 0:
+            out[n] = (float("nan"), float("nan"))
+            continue
+        w, b = fit_logreg(Xz[tr], ytr, steps=steps, wd=wd)
+        out[n] = (best_single_train_test(Xz[tr], ytr, Xz[te], yte), auroc_vec((Xz[te] @ w + b).float(), yte))
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply interpretability/sparse_autoencoders/recipes/evo2/src/evo2_sae/eval/probing/probe.py:117]: Done in `942071d8` — guard is now `yte.sum() in (0, len(yte))`, so all-positive test folds are skipped too (AUROC undefined).

File: interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    for nm in ("sae", "dense"):
+        if nm not in z.files:
+            continue
+        X = torch.from_numpy(z[nm]).to(dev).float()
+        Xz = (X - X.mean(0)) / (X.std(0) + 1e-6)
+        tr, te = split_indices(X.shape[0], a.test_frac, a.seed)
+        _, ca, _ = decode_eval(Xz[tr], codon[tr], Xz[te], codon[te], ncod, steps=a.steps, wd=a.weight_decay)
+        _, aaa, _ = decode_eval(Xz[tr], aa[tr], Xz[te], aa[te], naa, steps=a.steps, wd=a.weight_decay)
+        trn = torch.from_numpy(np.nonzero(~np.isin(codon_np, hidx))[0]).to(dev)
+        W, b = fit_softmax(Xz[trn], aa[trn], naa, steps=a.steps, wd=a.weight_decay)
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py]: Done in `942071d8` — codon-aa now standardizes via `_z(X, tr)` (train-split stats only); no more test-set leakage.

File: bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+def _add_model_args(p, *, required=(), max_tokens=160_000):
+    """Shared model + encoding args for the model-backed subcommands (extract/euk-f1/domain-eval)."""
+    for arg in ("--evo2-ckpt-dir", "--sae-checkpoint", "--fasta", *required):
+        p.add_argument(arg, required=True)
+    p.add_argument("--layer", type=int, required=True)
+    p.add_argument("--max-tokens", type=int, default=max_tokens)
+    p.add_argument("--seq-len", type=int, default=1024)
+    p.add_argument("--batch-size", type=int, default=8)
+    p.add_argument("--auroc-device", default="cuda:1")
```
[the remark this answers is not in the document]
Comment (polinabinder1) [reply bionemo-recipes/interpretability/sparse_autoencoders/recipes/evo2/scripts/probe.py]: Done in `942071d8` — `--auroc-device` defaults to `None` and falls back to `--device`; no hardcoded `cuda:1`.
